### PR TITLE
Avoid running QEventLoop on main thread (in WFS provider)

### PR DESCRIPTION
**Pull request by m-kuhn**
This is an alternative approach to running QEventLoops on the main thread.
In case we need an event loop and we are on the main thread, we just run it in a new thread and wait for that thread to be finished.

It would of course be preferable not to block the main thread while we download something, but if that's not (easily) possible this approach is hopefully more stable than running the event loop on the main thread, which I think is responsible for the crashes below.

For reviewers, the [feature counter commit](https://github.com/qgis/QGIS/pull/7055/commits/75d3ff282dfdc79331fa2515efdd300d2a0561c7) is a minimal example of the concept. The other commit just wraps more code in more or less the same approach.

Linux:

```
0x00007fe5b6f62a56 in socketNotifierSourceCheck(_GSource*) () from /lib64/libQt5Core.so.5
[Current thread is 1 (Thread 0x7fe5c3902400 (LWP 6195))]
#0  0x00007fe5b6f62a56 in socketNotifierSourceCheck(_GSource*) () at /lib64/libQt5Core.so.5
#1  0x00007fe5b0e368c9 in g_main_context_check (context=context@entry=0x1618610, max_priority=0, fds=fds@entry=0x1bf2470, n_fds=n_fds@entry=7) at gmain.c:3701
        result = <optimized out>
        check = 0x7fe5b6f62a00 <socketNotifierSourceCheck(_GSource*)>
        source = 0x1739470
        iter = {context = 0x1618610, may_modify = 1, current_list = 0x7fe574003500 = {0x7fe5740034e0, 0x17da0a0}, source = 0x1739470}
        pollrec = <optimized out>
        n_ready = 1
        i = <optimized out>
#2  0x00007fe5b0e36e40 in g_main_context_iterate (context=context@entry=0x1618610, block=block@entry=1, dispatch=dispatch@entry=1, self=<optimized out>) at gmain.c:3865
        max_priority = 2147483647
        timeout = 14
        some_ready = <optimized out>
        nfds = 7
        allocated_nfds = 7
        fds = 0x1bf2470
#3  0x00007fe5b0e36fac in g_main_context_iteration (context=0x1618610, may_block=1) at gmain.c:3929
        retval = <optimized out>
#4  0x00007fe5b6f62c2f in QEventDispatcherGlib::processEvents(QFlags<QEventLoop::ProcessEventsFlag>) () at /lib64/libQt5Core.so.5
#5  0x00007fe5b6f1096a in QEventLoop::exec(QFlags<QEventLoop::ProcessEventsFlag>) () at /lib64/libQt5Core.so.5
#6  0x00007fe55bc85684 in QgsWfsRequest::sendGET(QUrl const&, bool, bool, bool) (this=0x7fff9966dc58, url=..., synchronous=true, forceRefresh=false, cache=true) at /home/mku/dev/qgis/QGIS/src/providers/wfs/qgswfsrequest.cpp:146
        modifiedUrl = {d = 0x4516cd0}
        request = {d = {d = 0x4104340}}
        loop = <incomplete type>
#7  0x00007fe55bc9b098 in QgsWFSFeatureHitsRequest::getFeatureCount(QString const&, QString const&) (this=0x7fff9966dc58, WFSVersion=..., filter=...) at /home/mku/dev/qgis/QGIS/src/providers/wfs/qgswfsshareddata.cpp:1192
        getFeatureUrl = {d = 0x4516cd0}
        buffer = @0x20a5ee0: {d = 0x7fe5c25c9160 <vtable for QgsLayerTreeMapCanvasBridge+16>}
        error = {static null = {<No data fields>}, d = 0x1e9e450}
        domDoc = {<QDomNode> = {impl = 0x0}, <No data fields>}
        doc = {<QDomNode> = {impl = 0x7fff9966dc58}, <No data fields>}
        numberOfFeatures = {static null = {<No data fields>}, d = 0x0}
#8  0x00007fe55bc9abe9 in QgsWFSSharedData::getFeatureCount(bool) (this=0x44891d0, issueRequestIfNeeded=true) at /home/mku/dev/qgis/QGIS/src/providers/wfs/qgswfsshareddata.cpp:1106
        request = {<QgsWfsRequest> = {<QObject> = {<No data fields>}, static staticMetaObject = {d = {superdata = 0x7fe5b7356280 <QObject::staticMetaObject>, stringdata = 0x7fe55bcc6b38 <qt_meta_stringdata_QgsWfsRequest>, data = 0x7fe55bcc6a30 <qt_meta_data_QgsWfsRequest>, static_metacall = 0x7fe55bcbbc00 <QgsWfsRequest::qt_static_metacall(QObject*, QMetaObject::Call, int, void**)>, relatedMetaObjects = 0x0, extradata = 0x0}}, mUri = {mURI = {static staticMetaObject = {d = {superdata = 0x0, stringdata = 0x7fe5bfedde40 <qt_meta_stringdata_QgsDataSourceUri>, data = 0x7fe5bfedddc0 <qt_meta_data_QgsDataSourceUri>, static_metacall = 0x0, relatedMetaObjects = 0x0, extradata = 0x0}}, mHost = {static null = {<No data fields>}, d = 0x7fe5b6fac7c0 <QArrayData::shared_null>}, mPort = {static null = {<No data fields>}, d = 0x7fe5b6fac7c0 <QArrayData::shared_null>}, mDriver = {static null = {<No data fields>}, d = 0x7fe5b6fac7c0 <QArrayData::shared_null>}, mService = {static null = {<No data fields>}, d = 0x7fe5b6fac7c0 <QArrayData::shared_null>}, mDatabase = {static null = {<No data fields>}, d = 0x7fe5b6fac7c0 <QArrayData::shared_null>}, mSchema = {static null = {<No data fields>}, d = 0x7fe5b6fac7c0 <QArrayData::shared_null>}, mTable = {static null = {<No data fields>}, d = 0x7fe5b6fac7c0 <QArrayData::shared_null>}, mGeometryColumn = {static null = {<No data fields>}, d = 0x7fe5b6fac7c0 <QArrayData::shared_null>}, mSql = {static null = {<No data fields>}, d = 0x7fe5b6fac760}, mAuthConfigId = {static null = {<No data fields>}, d = 0x7fe5b6fac7c0 <QArrayData::shared_null>}, mUsername = {static null = {<No data fields>}, d = 0x7fe5b6fac7c0 <QArrayData::shared_null>}, mPassword = {static null = {<No data fields>}, d = 0x7fe5b6fac7c0 <QArrayData::shared_null>}, mSSLmode = QgsDataSourceUri::SslPrefer, mKeyColumn = {static null = {<No data fields>}, d = 0x7fe5b6fac7c0 <QArrayData::shared_null>}, mUseEstimatedMetadata = false, mSelectAtIdDisabled = false, mWkbType = QgsWkbTypes::Unknown, mSrid = {static null = {<No data fields>}, d = 0x7fe5b6fac7c0 <QArrayData::shared_null>}, mParams = {d = 0x37bc890}}, mAuth = {mUserName = {static null = {<No data fields>}, d = 0x7fe5b6fac7c0 <QArrayData::shared_null>}, mPassword = {static null = {<No data fields>}, d = 0x7fe5b6fac7c0 <QArrayData::shared_null>}, mAuthCfg = {static null = {<No data fields>}, d = 0x7fe5b6fac7c0 <QArrayData::shared_null>}}, mGetEndpoints = {d = 0x442d9d0}, mPostEndpoints = {d = 0x442dae0}}, mReply = 0x361a6f0, mErrorMessage = {static null = {<No data fields>}, d = 0x7fe5b6fac7c0 <QArrayData::shared_null>}, mErrorCode = QgsWfsRequest::NoError, mResponse = {d = 0x7fe5b6fac7c0 <QArrayData::shared_null>}, mIsAborted = false, mForceRefresh = false, mTimedout = false, mGotNonEmptyResponse = false}, static staticMetaObject = {d = {superdata = 0x7fe55bef6e50 <QgsWfsRequest::staticMetaObject>, stringdata = 0x7fe55bcc6ff0 <qt_meta_stringdata_QgsWFSFeatureHitsRequest>, data = 0x7fe55bcc6e60 <qt_meta_data_QgsWFSFeatureHitsRequest>, static_metacall = 0x7fe55bcbc690 <QgsWFSFeatureHitsRequest::qt_static_metacall(QObject*, QMetaObject::Call, int, void**)>, relatedMetaObjects = 0x0, extradata = 0x0}}}
        featureCount = 32767
#9  0x00007fe55bc3c686 in QgsWFSProvider::featureCount() const (this=0x190b160) at /home/mku/dev/qgis/QGIS/src/providers/wfs/qgswfsprovider.cpp:729
...
```

or

Windows:

```
QIODevice::channelReadyRead :

QIODevice::write :

QNetworkReply::url :

QNetworkReply::uploadProgress :

QObject::event :

QApplicationPrivate::notify_helper :

QApplication::notify :

QgsApplication::notify qgsapplication.cpp:325

QCoreApplication::notifyInternal2 :

QCoreApplicationPrivate::sendPostedEvents :

QEventDispatcherWin32::processEvents :

TranslateMessageEx :

TranslateMessage :

QEventDispatcherWin32::processEvents :

QEventLoop::exec :

QgsWFSFeatureDownloader::run qgswfsfeatureiterator.cpp:482

QgsWFSThreadedFeatureDownloader::run qgswfsfeatureiterator.cpp:804

QThread::start :

BaseThreadInitThunk :

RtlUserThreadStart :
```

### CODE REVIEW

**Comment by nyalldawson** [comment]
@wonder-sk a possible approach for our recent network request discussion?

File: src/providers/wfs/qgswfsfeatureiterator.cpp [hunk 1]
```diff
@@ -447,297 +450,320 @@ void QgsWFSFeatureDownloader::run( bool serializeFeatures, int maxFeatures )
...
+          {
+            // e.g: http://services.cuzk.cz/wfs/inspire-cp-wfs.asp?SERVICE=WFS&REQUEST=GetFeature&VERSION=2.0.0&TYPENAMES=cp:CadastralParcel
+            truncatedResponse = true;
+          }
+          break;
         }
       }
 
-      if ( finished )
+      parser.release();
```
Comment (wonder-sk) [inline src/providers/wfs/qgswfsfeatureiterator.cpp]: Doesn't this leak memory?

Comment (m-kuhn) [reply src/providers/wfs/qgswfsfeatureiterator.cpp]: thanks, good catch

**Comment by m-kuhn** [comment]
@rouault Sounds like a good plan, I'll have to look into that. Important is, that there is no way to invoke this in a bad way through the public API (and have asserts in place that prevent from using this in a bad way through non-public API).

For now I have only put the counter task in a separate thread, since it's an easier scope to test the proof of concept. How does that look to you?

**Comment by rouault** [comment]
> How does that look to you?

LGTM. Thanks for adressing that

**Comment by m-kuhn** [comment]
Is the auth mechanism also implemented with signal/slot connections that need to be executed synchroneously?

@elpaso do you have more information about that?

**Comment by m-kuhn** [comment]
I assume it is related to the `QgsNetworkAccessManager::authenticationRequired` signal which is sent over a `Qt::BlockingConnection` to the main thrad if it's not on the main thread. This leads to a deadlock.

I'm not completely sure, why authentication handling needs to be done on the main thread, but I assume it's potential user interaction (@elpaso can you confirm?)

If this is the case we'll need to fiddle on this approach which at least only runs `processEvents` if we actually get to the point of required user interaction (and keeps the other 99% of users crash-free).

If this is not the case and it's for some other reason, I've got a simpler commit lying around here.

**Comment by elpaso** [comment]
@m-kuhn yes: I think that (for HTTP basic auth only) there is a chance that the user must be prompted with the authentication dialog in order to enter the credentials during a network connection.

There is also a mechanism for the authentication `QgsCredentials` machinery to work on console applications and that is also handled with signals and slots from `QgisApp::namAuthenticationRequired`

**Comment by m-kuhn** [comment]
Thanks @elpaso 
that just got a whole lot more complicated :/

**Comment by m-kuhn** [comment]
I can't get those tests to fail locally, nor do I have any possibility to reproduce this locally. I have a tendency to believe that it's just travis bitching around here and to just merge this, in which case the following will happen:

 * 99% of all users with WFS will have less crashes (real world problem, various crashes reported)
 * 1% of all users with WFS who do use some login credentials popping up on loading of a WFS layer might get a freeze and shout out loud

Does anyone know how to access a WFS server, where QGIS will ask me for login credentials with a dialog?

@elpaso @dakcarto @rouault @luipir (and feel free to ping anyone else who could have an idea ;) )

**Comment by elpaso** [comment]
@m-kuhn you can use the server script in https://github.com/qgis/QGIS/blob/master/tests/src/python/qgis_wrapped_server.py passing in some env vars you can have HTTP Basic auth easily set up. 
From QGIS client - if you do not configure authentication in the WFS configuration dialog - a dialog should pop up when retrieving the capabilities.

This is also configured in CI test in https://github.com/qgis/QGIS/blob/master/tests/src/python/test_authmanager_password_ows.py#L98 

Honestly, I don't like the idea of blacklisting yet another auth test which is also an integration test for both client and server components of WFS, this kind of test helped to spot out real issues in the past.

I dont know if this could be related to this issue but be aware that the QGIS authmanager system is not thread safe, a work to refactor it completely in a thread-safe manner was initiated last year but I left the company that was sponsoring that effort (which is not trivial at all) before the work could be completed, see: https://github.com/boundlessgeo/QGIS/tree/bd-2648-auth-thread-safety-refactoring

**Comment by m-kuhn** [comment]
Thanks for the inputs, I'll have a look. I also don't like leaving this unprotected. But I also don't like the fact that the auth system is currently taking the rest of the system hostage by it's requirements regarding threads.

I see 2 options, where the second one which is targeted here has 2 options again. Was the auth-threading branch tackling the same problem?

 1. Never use blocking NAM requests from the main thread, side-effect: adds code complexity
 2. Spawn a thread and freeze the main thread if a blocking NAM request runs on main thread
   a. start auth dialogs in a new process
   b. in case there's an auth dialog, freeze the thread, return control to the main thread, run a local event loop (still dangerous!) for the auth dialog and then hand resume the networking thread

**Comment by m-kuhn** [comment]
Last call for a review, works fine with authentication now (tested, thanks a lot to @elpaso for the help in setting up a local authentication enabled wfs server) and hopefully improves stability with the QEventLoop in a separate thread.

This is mostly a proof of concept, if it turns out to live up to the expectations, it's still open to put encapsulate this logic into an easy to use and failproof class.

**Pull request merged**